src/script_interface/ScriptInterfaceBase.cpp

```diff
@@ -53,4 +64,75 @@ std::weak_ptr<ScriptInterfaceBase> &
 ScriptInterfaceBase::get_instance(ObjectId id) {
   return Utils::AutoObjectId<ScriptInterfaceBase>::get_instance(id);
 }
+
+/* Checkpointing functions. */
+
+/**
+ * @brief  Return a Variant representation of the state of the object.
+ *
+ * This should returne the internal state of the instance, so that
+ * the instance can be restored from this information.  The default
+ * implementation stores all the public parameters, including object
+ * parameters that are captured by calling get_state on them.
+ */
+Variant ScriptInterfaceBase::get_state() const {
+  std::vector<Variant> state;
+
+  auto params = this->get_parameters();
+  state.reserve(params.size());
+
+  for (auto const &p : params) {
+    state.push_back(std::vector<Variant>{
+        {p.first, boost::apply_visitor(Serializer{}, p.second)}});
+  }
+
+  return state;
+}
+
+void ScriptInterfaceBase::set_state(Variant const &state) {
+  using boost::get;
+  using std::vector;
+
+  VariantMap params;
+  UnSerializer u;
+
+  for (auto const &v : get<vector<Variant>>(state)) {
+    auto const &p = get<vector<Variant>>(v);
+    params[get<std::string>(p.at(0))] = boost::apply_visitor(u, p.at(1));
+  }
+
+  this->construct(params);
+}
+
+/**
+ * @brief Returns a binary representation of the state often
+ *        the instance, as returned by @f get_state().
+ */
+std::string ScriptInterfaceBase::serialize() const {
+  std::stringstream ss;
+  boost::archive::binary_oarchive oa(ss);
+  auto v = Serializer{}(this->id());
+
+  oa << v;
+  return ss.str();
+}
+
+/**
+ * @brief Creates a new instance from a binary state,
+ *        as returned by @f serialize().
+ */
+std::shared_ptr<ScriptInterfaceBase>
+ScriptInterfaceBase::unserialize(std::string const &state) {
+  std::stringstream ss(state);
```

Comment:
You could use
```cpp
boost::iostreams::array_source asource(state.c_str(), state.size());
boost::iostreams::stream<boost::iostreams::array_source> ss(asource);
boost::archive::binary_iarchive ar(ss);
```
to prevent copying `state`.  Needs
```cpp
#include <boost/iostreams/device/array.hpp>
#include <boost/iostreams/stream.hpp>
```

Reply:
That's a better solution. It would be nice to have zero-copy also for the output direction (::serialize()), but that seems not possible with either std or boost streams.